### ISSUE TITLE
Thanks for your excellent work! How can I train and test on my own dataset use Dave?

### ISSUE THREAD

**Issue by changcongxun** [jerpelhan/DAVE#12]

**Comment by GioFic95** [comment]
I am currently working on training and testing on my own dataset and I found some issues while adapting the code.

## In `train_similarity.py`

1. The logged val loss is not divided by the length of the validation set
https://github.com/jerpelhan/DAVE/blob/7266874a0c2714a8926e0db17bf547a369ac95de/train_similarity.py#L208
while the train loss is divided by the length of the training set
https://github.com/jerpelhan/DAVE/blob/7266874a0c2714a8926e0db17bf547a369ac95de/train_similarity.py#L207
Is there a reason for this?

## In `utils/data.py`

2. In the `FSC147WithDensityMapSimilarityStitched` dataset, used for the second stage of the training, images are sampled from the annotation file (which seems to be the same for all the splits)
https://github.com/jerpelhan/DAVE/blob/7266874a0c2714a8926e0db17bf547a369ac95de/utils/data.py#L528
instead of from the `images_names` list obtained from the split file as in the `FSC147WithDensityMapDOWNSIZE` dataset, used for the first stage of the training
https://github.com/jerpelhan/DAVE/blob/7266874a0c2714a8926e0db17bf547a369ac95de/utils/data.py#L154
If I'm not mistaken, in this way train images could be used in validation. And, if the annotation file is not sorted with all the training images at the beginning, test images could be used during training.

3. For an analogous reason, i.e. to avoid test images being seen during training, shouldn't the `candidate_images` for stitching come from the same subset as the "main" image?
https://github.com/jerpelhan/DAVE/blob/7266874a0c2714a8926e0db17bf547a369ac95de/utils/data.py#L555
What I mean is adding a second condition to this `if` statement:
```python
candidate_images = [im for im, cl in self.classes.items() if cl != img_class and im in self.image_names]
```

4. The `new_density_map` is defined but never used
https://github.com/jerpelhan/DAVE/blob/7266874a0c2714a8926e0db17bf547a369ac95de/utils/data.py#L575
It should be ok because we only need the number of objects in the "main" image, but it may be confusing.

Please let me know what you think about these observations.

**Comment by abhishekva65** [comment]
have you tried yet?

**Comment by jerpelhan** [comment]
To train DAVE on your own dataset, follow these steps:

1. Train LOCA: First, train the LOCA model using the instructions provided on the GitHub [here](https://github.com/djukicn/loca) with your data.
2. Run train_det.sh: Use the trained LOCA model and run train_det.sh, updating the dataloaders to accommodate your dataset. The verification network configuration can likely remain unchanged.

Could you provide more details about your dataset? Specifically, is it significantly different from the dataset used in the original DAVE training?

**Comment by abhishekva65** [comment]
Yes, our dataset is significantly different from the one used in DAVE. Our objective is to classify and count the number of items per class in an image. For example, we want to count the number of pencils of each color in a single image:

Yellow: 7
Red: 6
Green: 6
Blue: 7
Would it be feasible to achieve this using DAVE with necessary modifications to accommodate our specific use case? One possible approach we considered is running DAVE multiple times, each time using an exemplar belonging to a different class. Do you think this method could work, or would you suggest an alternative solution?

<img width="528" alt="image" src="https://github.com/user-attachments/assets/8cdef895-b5ab-4f98-aa10-e643060f79e5">

**Comment by abhishekva65** [comment]
the example i have given is fairly easy to differentiate but actual product we are using has only a slight difference in its classes.
so can we adjust the similarity score?

**Comment by jerpelhan** [comment]
Yes, that's a different task. In FSC, the goal is to count all objects of the same class (e.g., pencils), not by color. I recommend running DAVE as usual with exemplars (including multiple colors of pencils) and then tweaking the verification (clustering) part. Instead of clustering based on exemplars, try different thresholds (similarity score) to see if it clusters objects by color. 
If the default version doesn’t manage to do this successfully, you may need to retrain the verification part of DAVE (feature projection).
